atomic_reactor/plugins/pre_koji_delegate.py

```diff
@@ -0,0 +1,124 @@
+        self.osbs = None
+
+    def cancel_build(self):
+        """
+        cancel current build
+        """
+        build_name = self.metadata.get("name")
+        if build_name:
+            self.osbs.cancel_build(build_name)
+
+    def delegate_task(self):
+        """
+        create new koji task for autorebuild
+        """
+        user_params = os.environ['USER_PARAMS']
+        user_data = json.loads(user_params)
+
+        git_uri = user_data.get('git_uri')
+        git_ref = user_data.get('git_ref')
+        source = "%s#%s" % (git_uri, git_ref)
+        container_target = user_data.get('koji_target')
+
+        koji_task_id = self.metadata.get('labels', {}).get('original-koji-task-id')
+        if not koji_task_id:
+            koji_task_id = self.metadata.get('labels', {}).get('koji-task-id')
+
+        task_opts = {}
+        for key in ('yum_repourls', 'git_branch', 'signing_intent', 'compose_ids'):
+            if key in user_data:
+                if user_data[key]:
+                    task_opts[key] = user_data[key]
+        task_opts['triggered_after_koji_task'] = int(koji_task_id)
+
+        task_id = self.kojisession.buildContainer(source, container_target, task_opts,
+                                                  priority=self.task_priority)
+
+        self.log.info('Created intermediate task: %s', task_id)
+
+    def run(self):
+        """
+        run the plugin
+        """
+        if self.delegate_enabled:
+            self.workflow.triggered_after_koji_task = self.triggered_after_koji_task
+
+        koji_task_id = self.metadata.get('labels', {}).get('koji-task-id')
+        task_info = self.kojisession.getTaskInfo(koji_task_id, request=True)
+        task_running = koji.TASK_STATES[task_info['state']] == 'OPEN'
+
+        # we don't want to plugin continue when:
+        # delegate_task isn't enabled
+        # build isn't autorebuild
+        # triggered_after_koji_task was provided, but task is running,
+        # reason for this is, when we once enable delegating, after first autorebuild
+        # buildConfig will already have triggered_after_koji_task in user_params
+        # so when koji-task-id for build is running task, that means it is that new
+        # already delegated task
+        if (not self.delegate_enabled or (self.triggered_after_koji_task and task_running) or
+                not is_rebuild(self.workflow)):
+            self.log.info("Not delegating build")
+            return
+
+        self.osbs = get_openshift_session(self.workflow, NO_FALLBACK)
+
+        self.delegate_task()
+        self.cancel_build()
+        self.log.info('Delegated build, will stop here')
+        # we don't want to others plugin to continue, and
+        # canceling build might take some time
+        # if we would raise error, exit plugins including sendmail would run,
+        # but we don't want to send email about failed build, because it isn't "real" build
+        time.sleep(3000)
```

Comment:
Oh, or this will be just killed during sleep?

Reply:
yes it will be killed during sleep, because we don't want to run any other plugins, not even exit plugins, but cancel can take a while

Reply:
Should this be `signal.pause()`?

Reply:
I'll actually delete all exit plugins, and raise canceled build exception